politeiad/backend/gitbe/gitbe.go

```diff
@@ -1021,78 +1186,279 @@ func (g *gitBackEnd) New(metadata string, files []backend.File) (*backend.Record
 		return nil, err
 	}
 
-	// git checkout -b id
-	err = g.gitNewBranch(g.unvetted, id)
+	var errReturn error
+	brm, err := g.newRecord(token, metadata, fa)
+	if err != nil {
+		// git stash
+		err2 := g.gitStash(g.unvetted)
+		if err2 != nil {
+			// We are in trouble!  Consider a panic.
+			return nil, err2
+		}
+
+		brm = nil
+		errReturn = err
+	}
+
+	// git checkout master
+	err = g.gitCheckout(g.unvetted, "master")
 	if err != nil {
 		return nil, err
 	}
 
-	// Process files.
-	path := filepath.Join(g.unvetted, id, defaultPayloadDir)
-	err = os.MkdirAll(path, 0774)
+	return brm, errReturn
+}
+
+// update takes various parameters to update a record.  Note that this function
+// must be wrapped by a function that delivers the call with the unvetted repo
+// sitting in master.  The idea is that if this function fails we can simply
+// unwind it by calling a git stash.
+// Function must be called with the lock held.
+func (g *gitBackEnd) update(token []byte, mdAppend []backend.MetadataStream, mdOverwrite []backend.MetadataStream, fa []file, filesDel []string) (*backend.RecordMetadata, error) {
+	id := hex.EncodeToString(token)
+
+	// See if branch already exists
+	branches, err := g.gitBranches(g.unvetted)
 	if err != nil {
 		return nil, err
 	}
+	var found bool
+	for _, v := range branches {
+		if !util.IsDigest(v) {
+			continue
+		}
+		if v == id {
+			found = true
+			break
+		}
+	}
 
-	hashes := make([]*[sha256.Size]byte, 0, len(fa))
+	if found {
+		// Branch exists, modify branch
+		err := g.gitCheckout(g.unvetted, id)
+		if err != nil {
+			return nil, backend.ErrRecordNotFound
+		}
+	} else {
+		// Branch does not exist, create it if record exists
+		fi, err := os.Stat(filepath.Join(g.unvetted, id))
+		if err != nil {
+			if os.IsNotExist(err) {
+				return nil, backend.ErrRecordNotFound
+			}
+		}
+		if !fi.IsDir() {
+			return nil, fmt.Errorf("unvetted repo corrupt: %v "+
+				"is not a dir", fi.Name())
+		}
+		// git checkout -b id
+		err = g.gitNewBranch(g.unvetted, id)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// We now are sitting in branch id
+
+	// Load MD
+	log.Tracef("updating %x", token)
+	brm, err := loadMD(g.unvetted, id)
+	if err != nil {
+		return nil, err
+	}
+	if !((found && brm.Status == backend.MDStatusVetted) ||
+		brm.Status == backend.MDStatusUnvetted ||
+		brm.Status == backend.MDStatusIterationUnvetted) {
+		return nil, fmt.Errorf("can not update record that "+
+			"has status: %v", brm.Status)
+	}
+
+	// Verify all deletes before executing
+	for _, v := range filesDel {
+		fi, err := os.Stat(filepath.Join(g.unvetted, id,
+			defaultPayloadDir, v))
+		if err != nil {
+			if os.IsNotExist(err) {
```

Comment:
That is by design.  The record must exist.

Reply:
From the golang docs: "Stat returns a FileInfo describing the named file. If there is an error, it will be of type *PathError."

So we should be cool.